### ISSUE TITLE
How to ignore bad json inputs

### ISSUE THREAD

**Issue by percula** [Kotlin/kotlinx.serialization#184]
This is my serializable class

```kotlin
@Serializable
data class Location(@Optional val longlat: MutableList<Double>? = mutableListOf())
```

Most of my JSON is correct, but sometimes instead of containing a List<Double> or null, it contains a map. Is there a way to disregard this bad JSON input and continue deserializing the object? Currently it throws an exception when it gets to the map, saying it expected a list.

If it matters, currently I am using https://github.com/JakeWharton/retrofit2-kotlinx-serialization-converter with JSON.nonstrict.

**Comment by percula** [comment]
Thanks @sandwwraith . I agree, fixing the API is the best solution and I've already filed a bug with our backend team. However, it would be great if we could add a parameter to @Optional such as @Optional(catchExceptions = true). In this scenario, it would catch the exception, log it, and move on to the next value.

**Comment by pdvrieze** [comment]
@percula What you want is probably a case where you want a custom reader that extends from the default json reader, but adds the logging/etc. you want - there are too many ways you may want to handle it. Of course this assumes that the reader is sufficiently extensible to allow you to do this. This may be something to consider for the library.

**Comment by percula** [comment]
@pdvrieze , yes that would be very helpful. Right now it is a lot of work to create a custom reader or serializer. If we could extend the default, code-generated reader and serializer that would greatly simplify the process.

**Comment by pdvrieze** [comment]
@percula It is not the generated code you want to extend (that is possible as well, that is the serializer), it is the json reader that you want to extend: `MyJSON.parse(someJsonString)` instead of `JSON.parse(someJsonString)`.
